crates/api/src/community/multi_community_follow.rs

```diff
@@ -0,0 +1,91 @@
+use activitypub_federation::config::Data;
+use actix_web::web::Json;
+use lemmy_api_common::{
+  community::FollowMultiCommunity,
+  context::LemmyContext,
+  send_activity::{ActivityChannel, SendActivityData},
+  utils::{check_local_user_valid, community_follow_many, community_unfollow_many},
+  SuccessResponse,
+};
+use lemmy_db_schema::{
+  source::multi_community::{MultiCommunity, MultiCommunityFollowForm},
+  traits::Crud,
+};
+use lemmy_db_schema_file::enums::CommunityFollowerState;
+use lemmy_db_views_community::impls::CommunityQuery;
+use lemmy_db_views_local_user::LocalUserView;
+use lemmy_db_views_site::SiteView;
+use lemmy_utils::error::LemmyResult;
+
+pub async fn follow_multi_community(
+  data: Json<FollowMultiCommunity>,
+  context: Data<LemmyContext>,
+  local_user_view: LocalUserView,
+) -> LemmyResult<Json<SuccessResponse>> {
+  check_local_user_valid(&local_user_view)?;
+  let multi_community_id = data.multi_community_id;
+  let person_id = local_user_view.person.id;
+  let multi = MultiCommunity::read(&mut context.pool(), multi_community_id).await?;
+  let site = SiteView::read_local(&mut context.pool()).await?;
+  let communities = CommunityQuery {
+    local_user: Some(&local_user_view.local_user),
+    multi_community_id: Some(multi_community_id),
+    ..Default::default()
+  }
+  .list(&site.site, &mut context.pool())
+  .await?;
+
+  let follow_state = if multi.local {
+    CommunityFollowerState::Accepted
+  } else {
+    CommunityFollowerState::Pending
+  };
+  let form = MultiCommunityFollowForm {
+    multi_community_id,
+    person_id,
+    follow_state,
+  };
+
+  if data.follow {
+    MultiCommunity::follow(&mut context.pool(), &form).await?;
+
+    // Get all communities which are part of the multi and not yet followed by current user
+    let to_follow: Vec<_> = communities
+      .into_iter()
+      .filter(|c| {
+        let actions = c.community_actions.clone().unwrap_or_default();
+        actions.followed.is_none()
+      })
+      .map(|c| c.community)
+      .collect();
+
+    // Then follow them
+    community_follow_many(&local_user_view.person, &to_follow, &context).await?;
```

Comment:
The way it works on reddit, just because you follow a multi-community does not mean you've subscribed to every community inside it.

I spose since you can't guarantee that these communities have local subscribers, then this is unavoidable. Just wanted to point out this difference.

The only other way around this I can think of, is to keep a fake user just for remote subscribes.

Reply:
Good idea, that will make the logic a lot simpler. I should have thought of that before writing all this code.

Reply:
Implemented the schema for this now.